### PR TITLE
Fix inconsistency between Python and C http request parsers.

### LINKED ISSUE #4972: inconsistent query arg handling between python and C http request parsers with yarl 1.6.0
🐞 **Describe the bug**

If you upgrade to yarl 1.6.0 and are using the C HttpRequestParser class, your query arguments are not dequoted (by which I mean, request.query['foo'] is something like `%7B%22val%22: 3%7D` and not `{"val": 3}`). If you use the Python HttpRequestParser class, yarl 1.6.0 does not change the behaviour.

💡 **To Reproduce**

```
>>> from aiohttp import http_parser
>>> p = http_parser.HttpRequestParserC(1, None); p.feed_data(b'GET /test?b=%22val%22 HTTP/1.1\r\n\r\n')[0][0][0].url.query['b']
'%22val%22'
>>> p = http_parser.HttpRequestParserPy(1, None); p.feed_data(b'GET /test?b=%22val%22 HTTP/1.1\r\n\r\n')[0][0][0].url.query['b']
'"val"'
```

(with yarl 1.5.1 both return "val").

📋 **Your version of the Python**

3.6 and 3.8 at least -- don't think it matters

📋 **Your version of the aiohttp/yarl/multidict distributions**

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.6.2
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: Nikolay Kim
Author-email: fafhrd91@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.6/dist-packages
Requires: attrs, async-timeout, yarl, multidict, chardet, typing-extensions, idna-ssl
```
```console
$ python -m pip show multidict
Name: multidict
Version: 4.7.6
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.6/dist-packages
Requires: 
```
```console
$ python -m pip show yarl
Name: yarl
Version: 1.6.0
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/local/lib/python3.6/dist-packages
Requires: idna, multidict, typing-extensions
```

📋 **Additional context**

The reason for the inconsistency is that the Python parser calls URL(path) whereas the C parser calls URL.build(path=path) and the latter appears to have changed behaviour in 1.6.0.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
They both now correctly parse URLs containing pct-encoded sequences and work with yarl 1.6.0.

~Yarl >= 1.6 is now required.~

Closes #4972

### CODE REVIEW

**Comment by webknjaz** [comment]
@serhiy-storchaka looks like the CI failures are caused by this change.

File: aiohttp/web_urldispatcher.py [hunk 1]
```diff
@@ -1121,3 +1115,15 @@ def add_routes(self,
...
+def _quote_path(value: str) -> str:
+    return URL.build(path=value).raw_path
+
+
+def _unquote_path(value: str) -> str:
+    return URL.build(path=value, encoded=True).path
+
+
+def _requote_path(value: str) -> str:
+    return _quote_path(value).replace('%25', '%')
```
Comment (webknjaz) [inline aiohttp/web_urldispatcher.py]: Why is this necessary only for `%25`? Shouldn't it be more generic?

File: tests/test_http_parser.py [hunk 2]
```diff
@@ -787,6 +788,70 @@ def test_url_parse_non_strict_mode(parser) -> None:
     assert payload.is_eof()
 
 
+@pytest.mark.parametrize(
+    'uri,path,query,fragment',
```
Comment (webknjaz) [inline tests/test_http_parser.py]: I've recently discovered that you can make the params a bit more readable in pytest:
```suggestion
    ('uri', 'path', 'query', 'fragment'),
```

**Comment by serhiy-storchaka** [comment]
Now there is a question: do we want to support yarl < 1.6 or require yarl >= 1.6?

File: tests/test_http_parser.py [hunk 3]
```diff
@@ -787,6 +788,70 @@ def test_url_parse_non_strict_mode(parser) -> None:
...
+
+    assert msg.path == uri
+    assert msg.url == URL(uri)
+    assert msg.url.path == path
+    assert msg.url.query == query
+    assert msg.url.fragment == fragment
+
+
+def test_parse_uri_utf8(parser) -> None:
+    text = ('GET /путь?ключ=знач#фраг HTTP/1.1\r\n\r\n').encode()
```
Comment (webknjaz) [inline tests/test_http_parser.py:815]: How about using something from extended Unicode like `☃`?

File: aiohttp/web_urldispatcher.py (hunk 1 above)
Comment (webknjaz) [reply aiohttp/web_urldispatcher.py]: Fair enough

**Comment by webknjaz** [comment]
> Now there is a question: do we want to support yarl < 1.6 or require yarl >= 1.6?

I'm okay with doing `>= 1.6` for master (future 4.0) but the backport may be a bit more tricky: what if the end-users are stuck with old yarl for some reason? OTOH since it fixes a bug, it'd be okay to increase the required version if it's needed for the bugfix.

**Comment by webknjaz** [comment]
re: CI failures -- `test_close[pyloop]` is the only test that fails with Python 3.8 under Win/Mac but it also fails on master so it's unrelated to this PR.

File: tests/test_http_parser.py (hunk 3 above)
Comment (serhiy-storchaka) [reply tests/test_http_parser.py:815]: I actually copied an example from some test for `yarl`.

**Comment by mwhudson** [comment]
Any update on this?

**Comment by TBBle** [comment]
What about this fix required yarl 1.6? My understanding of the [yarl change](https://github.com/aio-libs/yarl/pull/503) is that if we just ensure we specify `encoded=True` for all calls of `URL.build` (and `with_xxx` and operator `/`, but I don't think they're used here...), then that would be compatible with yarl 1.5 as well.

The core issue as I understood it was that we were passing encoded elements to `URL.build`, which defaults `encoded=False`, and it worked before yarl 1.6 because it was ignoring the `%` and there were no _other_ characters that needed re-encoding (because we'd already encoded it).

File: aiohttp/web_urldispatcher.py [hunk 4]
```diff
@@ -1121,3 +1115,18 @@ def add_routes(self,
         for route_def in routes:
             registered_routes.extend(route_def.register(self))
         return registered_routes
+
+
+def _quote_path(value: str) -> str:
+    return URL.build(path=value).raw_path
```
Comment (TBBle) [inline aiohttp/web_urldispatcher.py]: It might be cleaner to explicitly pass `encoded=False` here, to make clear what we're trying to do, and contrast it with `_unquote_path` below.

File: aiohttp/web_urldispatcher.py [hunk 5]
```diff
@@ -1121,3 +1115,18 @@ def add_routes(self,
...
+
+
+def _unquote_path(value: str) -> str:
+    return URL.build(path=value, encoded=True).path
+
+
+def _requote_path(value: str) -> str:
```
Comment (TBBle) [inline aiohttp/web_urldispatcher.py:1168]: I'm struggling with the logic of this. Is this just reimplementing the yarl 1.5 behaviour when `URL.build(..., encoded=False)` is given an encoded string?

If so, it might be handy to document that, because it _looks_ like it quotes the path and then unquotes it for `%`, and even with that context, I'm having trouble parsing why we'd want to do that.

**Comment by serhiy-storchaka** [comment]
It now works also with yarl <1.6.

**Review by asvetlov — APPROVED** [review]
LGTM

**Comment by asvetlov** [comment]
Thanks, @serhiy-storchaka
